cpp/daal/src/algorithms/tsne/tsne_gradient_descent_avx512_impl.i

```diff
@@ -0,0 +1,326 @@
+
+                    for (IdxType i = 0; i < (range / 16) * 16; i += 16)
+                    {
+                        prefetch_index = start_index + i + prefetch_dist;
+                        if (prefetch_index < nnz) DAAL_PREFETCH_READ_T0(&mem._pos[col[prefetch_index] - 1]);
+
+                        __m512i vec_iCol = _mm512_sub_epi32(_mm512_loadu_epi32((__m512i *)&col[start_index + i]), vec_1i);
+
+                        __m512 vec_point_xd = _mm512_sub_ps(vec_point_x, _mm512_i32gather_ps(vec_iCol, &mem._pos[0].x, 8));
+                        __m512 vec_point_yd = _mm512_sub_ps(vec_point_y, _mm512_i32gather_ps(vec_iCol, &mem._pos[0].y, 8));
+
+                        __m512 vec_pq = _mm512_div_ps(
+                            _mm512_loadu_ps((__m512 *)&val[start_index + i]),
+                            _mm512_add_ps(_mm512_fmadd_ps(vec_point_xd, vec_point_xd, _mm512_mul_ps(vec_point_yd, vec_point_yd)), vec_1));
+
+                        vec_point_xs = _mm512_fmadd_ps(vec_point_xd, vec_pq, vec_point_xs);
+                        vec_point_ys = _mm512_fmadd_ps(vec_point_yd, vec_pq, vec_point_ys);
+                    }
+
+                    mem._attr[iRow].x += _mm512_reduce_add_ps(vec_point_xs);
+                    mem._attr[iRow].y += _mm512_reduce_add_ps(vec_point_ys);
+
+                    for (IdxType i = (range / 16) * 16; i < range; ++i)
+                    {
+                        prefetch_index = start_index + i + prefetch_dist;
+                        if (prefetch_index < nnz) DAAL_PREFETCH_READ_T0(&mem._pos[col[prefetch_index] - 1]);
+
+                        iCol = col[start_index + i] - 1;
+
+                        y1d = row_point.x - mem._pos[iCol].x;
+                        y2d = row_point.y - mem._pos[iCol].y;
+
+                        sqDist = 1.0 + y1d * y1d + y2d * y2d;
+                        PQ     = val[start_index + i] / sqDist;
+
+                        mem._attr[iRow].x += PQ * y1d;
+                        mem._attr[iRow].y += PQ * y2d;
+                    }
+                }
+                else
+                {
+                    for (size_t index = row[iRow] - 1; index < row[iRow + 1] - 1; ++index)
+                    {
+                        prefetch_index = index + prefetch_dist;
+                        if (prefetch_index < nnz) DAAL_PREFETCH_READ_T0(&mem._pos[col[prefetch_index] - 1]);
+                        iCol = col[index] - 1;
+
+                        y1d    = row_point.x - mem._pos[iCol].x;
+                        y2d    = row_point.y - mem._pos[iCol].y;
+                        sqDist = services::internal::max<avx512, float>(float(0), y1d * y1d + y2d * y2d);
+                        PQ     = val[index] / (sqDist + 1.);
+
+                        // Apply forces
+                        mem._attr[iRow].x += PQ * y1d;
+                        mem._attr[iRow].y += PQ * y2d;
+
+                        logLocal[iSize++] = val[index] * multiplier * (1. + sqDist);
```

Comment:
You are using implicit conversion to double and then to float. 
```suggestion
                        logLocal[iSize++] = val[index] * multiplier * (1.f + sqDist);
```

Reply:
Updated the code.